### PR TITLE
Adding Collected Recipients and Trusted Senders

### PR DESCRIPTION
**Pull request by Shadow243**
Both the "Collected Recipients" and "Trusted Senders" address books are highly configurable. Users have the flexibility to link these address books to other writable address books, such as their primary address book. Moreover, users have the option to disable this functionality if it does not align with their preferences.

For users of Cypth, a configuration option has been introduced: "Automatically add outgoing email addresses to allow auto collection of contact." This further streamlines the process of building and maintaining contact lists, ensuring a seamless and efficient experience when using Roundcube's webmail service.

[https://kolabian.wordpress.com/2020/09/26/collected-recipients-and-trusted-senders/](https://kolabian.wordpress.com/2020/09/26/collected-recipients-and-trusted-senders/)

### CODE REVIEW

**Comment by kroky** [comment]
@Shadow243 please check your indentation to be consistent with the files you edit - you have used tabs instead of spaces.

**Comment by Shadow243** [comment]
Indentation corrected.

File: modules/imap/handler_modules.php [hunk 1]
```diff
@@ -1906,6 +1906,14 @@ public function process() {
                     $this->out('msg_download_args', $download_args.'&amp;imap_download_message=1');
                     $this->out('msg_attachment_remove_args', $download_args.'&amp;imap_remove_attachment=1');
                     $this->out('msg_show_args', sprintf("page=message&amp;uid=%s&amp;list_path=imap_%s_%s&amp;imap_show_message=1", $form['imap_msg_uid'], $form['imap_server_id'], $form['folder']));
+					
+                    if ($this->get('imap_allow_images', false)) {
+                        if ($this->module_is_supported('contacts') && $this->user_config->get('contact_auto_collect_setting', false)) {
+                            $this->out('collect_contacts', true);
+                            $this->out('collected_contact_email', $msg_headers["Return-Path"]);
+                            $this->out('collected_contact_name', $msg_headers["From"]);
+						}
```
Comment (kroky) [inline modules/imap/handler_modules.php:1915]: Indentation here is wrong.

File: modules/contacts/modules.php [hunk 2]
```diff
@@ -6,328 +6,510 @@
...
-            $contacts = $this->get('contact_store');
-            $contacts->sort('email_address');
-            $results = array_slice($contacts->search(array(
-                'display_name' => $val,
-                'email_address' => $val
-            )), 0, 10);
-        }
-        $this->out('contact_suggestions', $results);
-    }
+class Hm_Handler_process_contact_auto_collect_setting extends Hm_Handler_Module
```
Comment (kroky) [inline modules/contacts/modules.php:18]: Please revert all these syntax changes - I think you used an IDE to do it but this has negative side effects:
- makes it rather hard to review the actual code that changed
- introduces code changes in git history that pollute the history and don't improve anything

I'd like to see a diff of the actual changes you did here, so please either do them again on a branch starting from fresh master or revert the syntax changes by your IDE.

Comment (kroky) [reply modules/contacts/modules.php:18]: Alright, please let me know when you leave only the actual changes in this PR and I will review it again.

Comment (Shadow243) [reply modules/contacts/modules.php:18]: Indentation is Good now from: [https://github.com/cypht-org/cypht/commit/9a14becd4ed6616e319fa8e744e4b333ed6a654a](https://github.com/cypht-org/cypht/commit/9a14becd4ed6616e319fa8e744e4b333ed6a654a )

Comment (kroky) [reply modules/contacts/modules.php:18]: It is not only the indentation. There are hundreds of places where opening/closing brackets were moved to a new line. See diff here: https://github.com/cypht-org/cypht/pull/816/files

Comment (Shadow243) [reply modules/contacts/modules.php]: Oh you are right, I deleted the branch and recreated it.

File: modules/contacts/site.css [hunk 3]
```diff
@@ -35,3 +35,141 @@
 .mobile .contact_list_title, .mobile .add_contact { display: none; }
 .mobile .contact_fld { display: none; } .contact_src { display: none; }
 .mobile .add_contact_row { display: none; }
+
+.pcss3t { margin: 0;padding: 0;border: 0;outline: none;font-size: 0;text-align: left;padding-top: 15px; }
```
Comment (kroky) [inline modules/contacts/site.css:39]: This block seems like an external code/lib - can we put it in third_party and include separately?

File: modules/local_contacts/modules.php [hunk 4]
```diff
@@ -23,7 +23,7 @@ public function process() {
             if (!empty($addresses)) {
                 $contacts = $this->get('contact_store');
                 foreach ($addresses as $vals) {
-                    $contacts->add_contact(array('source' => 'local', 'email_address' => $vals['email'], 'display_name' => $vals['name']));
+                    $contacts->add_contact(array('source' => 'local', 'email_address' => $vals['email'], 'display_name' => $vals['name'], 'group' => isset($vals['contact_group']) ? $vals['contact_group'] : 'Personal Addresses'));
```
Comment (kroky) [inline modules/local_contacts/modules.php:26]: We have 3 more contact stores - carddav, ldap, gmail. Shouldn't we extend those as well to support new functionality?

File: modules/contacts/modules.php (hunk 2 above)
Comment (kroky) [reply modules/contacts/modules.php]: Thank you!

**Comment by kroky** [comment]
Please also rebase master and resolve conflicts.

File: modules/contacts/site.css (hunk 3 above)
Comment (Shadow243) [reply modules/contacts/site.css]: Done.

File: modules/local_contacts/modules.php (hunk 4 above)
Comment (Shadow243) [reply modules/local_contacts/modules.php:26]: Working on it.

**Comment by Shadow243** [comment]
> Please also rebase master and resolve conflicts.

Done.

File: modules/local_contacts/modules.php (hunk 4 above)
Comment (Shadow243) [reply modules/local_contacts/modules.php:26]: @kroky While waiting to have the identifiers to test other contact stores: carddav, ldap, gmail, can you merge this one and we will then do another PR for the rest ?

Comment (kroky) [reply modules/local_contacts/modules.php:26]: I am fine with separate PRs for the rest.

File: modules/contacts/modules.php [hunk 5]
```diff
@@ -10,6 +10,70 @@
...
+}
+
+/**
+ * @subpackage contacts/handler
+ */
+class Hm_Handler_store_contact_message extends Hm_Handler_Module {
+    public function process() {
+        if ($this->get('collect_contacts', false)) {
+            $addresses = process_address_fld($this->request->post['compose_to']);
+            $contact_list = $this->user_config->get('contacts', []);
```
Comment (kroky) [inline modules/contacts/modules.php:32]: Latest master has moved to a Repository pattern for storing and retrieving contacts. Please check how it is done and use appropriate code to get contacts - they are already initialized in one of the handler methods.

File: modules/contacts/modules.php [hunk 6]
```diff
@@ -10,6 +10,70 @@
...
+/**
+ * @subpackage contacts/handler
+ */
+class Hm_Handler_store_contact_allow_images extends Hm_Handler_Module {
+    public function process() {
+        if ($this->get('imap_allow_images', false) && $this->get('collect_contacts', false)) {
+            $email = str_replace(['<', '>'], '', $this->get('collected_contact_email', ''));
+            $name = $this->get('collected_contact_name', '');
+            $contacts = $this->get('contact_store');
+            $contact_list = $this->user_config->get('contacts', []);
```
Comment (kroky) [inline modules/contacts/modules.php:65]: Please use the repository to get the contacts.

File: modules/contacts/modules.php [hunk 7]
```diff
@@ -10,6 +10,70 @@
...
+            $existingEmails = array_column($contact_list, 'email_address');
+            if (!in_array($email, $existingEmails)) {
+                $contacts->add_contact(['source' => 'local', 'email_address' => $email, 'display_name' => $name, 'group' => $this->trans('Trusted Senders')]);
+                $this->user_config->set('contacts', $contacts->export());
```
Comment (kroky) [inline modules/contacts/modules.php:69]: Fix storing please.

**Comment by Shadow243** [comment]
Requested updates submitted/

**Pull request merged**